File: program_topic_go/others/order/shell_sort.go
package main

func shell_sort(arr []int) {
	l := len(arr)
	if l == 0 || arr == nil {
		return
	}

	for g := l / 2; g > 0; g /= 2 {
		for i := g; i < l; i++ {
			k, j := arr[i], i
			for j >= g && arr[j-g] > k {
				arr[j] = arr[j-g]
				j -= g
			}
			arr[j] = k
		}
	}
}

func main() {
	arr := []int{7, 0, 1, 9, 4, 6, 3, 5, 2, 8}
	print_arr(arr)
	shell_sort(arr)
	print_arr(arr)
}

//func print_arr(arr []int) {
//	l := len(arr)
//	if arr == nil || l == 0 {
//		return
//	}
//
//	for i := 0; i < l-1; i++ {
//		fmt.Print(arr[i], "\t")
//	}
//
//	fmt.Println(arr[l-1])
//}

File: program_topic_java/src/sort_algorithm/ShellSort.java
package sort_algorithm;

/*
希尔排序
希尔排序是把记录按下标的一定增量分组，对每组使用直接插入排序算法排序，随着增量逐渐减少，
每组包含的关键词越来越多，当增量减至1时，整个数组恰被分成一组，算法便终止。
 */
public class ShellSort {
    public static void main(String[] args) {
        int[] arr = {8, 0, 5, 7, 9, 1, 4, 2, 6, 3};
        ShellSort shellSort = new ShellSort();
        shellSort.printArray(arr);
        shellSort.shellSort(arr);
        shellSort.printArray(arr);
    }

    public void shellSort(int[] arr) {
        int size = arr.length;
        if (size == 0 || arr == null) {
            return;
        }

        for (int group = size / 2; group > 0; group /= 2) {
            for (int i = group; i < size; ++i) {
                int k = arr[i];
                int j = i;
                while (j >= group && arr[j - group] > k) {
                    arr[j] = arr[j - group];
                    j -= group;
                }
                arr[j] = k;
            }
        }
    }

    public void printArray(int[] arr) {
        int size = arr.length;
        if (size == 0 || arr == null) {
            return;
        }

        for (int i = 0; i < size - 1; ++i) {
            System.out.print(arr[i] + "\t");
        }
        System.out.println(arr[size - 1]);
    }
}
